File: Array/zhangjianan/20200703环路检测/detectCycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *detectCycle(struct ListNode *head) {
      if(head==NULL||head->next==NULL) return NULL;
  struct ListNode *slow=head;
  struct ListNode *fast=head;
  while(fast!=NULL&&fast->next!=NULL
  ){
      fast=fast->next->next;
      slow=slow->next;
      if(
          slow==fast
      )break;
  }
  if(fast!=slow)return NULL;
  slow=head;
  while(slow!=fast){
      slow=slow->next;
      fast=fast->next;
  }
  return slow;
}

File: Array/huangxinyu/kotlin/src/DetectCycle.kt
import com.ryujin.algorithm.ListNode

class DetectCycle {
    fun detectCycle(head: ListNode): ListNode? {
        var slow = head;
        var fast = head;
        while (fast.next != null) {
            fast = fast.next!!.next!!
            slow = slow.next!!
            if (slow == fast) {
                break
            }

        }
        if (slow != fast) {
            return null
        }
        fast = head
        while (fast != slow) {
            fast = fast.next!!
            slow = slow.next!!
        }
        return fast
    }
}